libyul/optimiser/SSATransform.cpp

```diff
@@ -40,23 +42,69 @@ void SSATransform::operator()(Identifier& _identifier)
 
 void SSATransform::operator()(ForLoop& _for)
 {
-	// This will clear the current value in case of a reassignment inside the
-	// init part, although the new variable would still be in scope inside the whole loop.
-	// This small inefficiency is fine if we move the pre part of all for loops out
-	// of the for loop.
-	(*this)(_for.pre);
+	// Avoid dealing with complicated scoping rule
+	assertThrow(_for.pre.statements.empty(), OptimizerException, "");
+
+	auto loc = _for.location;
+	// creates a VariableDeclaration statement that assigns _varName to new SSA var
+	// and update m_currentVariableValues
+	auto newVariableDecl = [&](YulString _varName) -> Statement
+	{
+		YulString newName = m_nameDispenser.newName(_varName);
+		m_currentVariableValues[_varName] = newName;
+		return
+			VariableDeclaration{
+				loc,
+				{TypedName{loc, newName, {}}},
+				make_unique<Expression>(Identifier{loc, _varName})
+			};
+	};
+	// prepend statements in _a to the beginning of _b, also empties _a
+	auto prependStatements = [](vector<Statement>& _a, vector<Statement>& _b)
+	{
+		std::move(_b.begin(), _b.end(), std::back_inserter(_a));
+		swap(_a, _b);
+		_a.clear();
+	};
+
+	// If existing variables get re-assigned in loop's body block or post block,
+	// we need to declare new variables to hold its value at the location where
+	// control flow joins, i.e beginning of post block and beginning of body block.
 
 	Assignments assignments;
 	assignments(_for.body);
 	assignments(_for.post);
+
+	// stores variables declared outside the loop but re-assigned in post/body block.
+	vector<YulString> varsToDeclare;
+	for (YulString varName: assignments.names())
+		if (m_currentVariableValues.count(varName))
+			varsToDeclare.push_back(varName);
+
 	for (auto const& var: assignments.names())
 		m_currentVariableValues.erase(var);
-
 	visit(*_for.condition);
+
+	// temporarily holds declarations to be prepend to the block
+	vector<Statement> decls;
+
+	for (auto const& var: assignments.names())
+		m_currentVariableValues.erase(var);
+	for (YulString varName: varsToDeclare)
+		decls.emplace_back(newVariableDecl(varName));
 	(*this)(_for.body);
+	prependStatements(decls, _for.body.statements);
```

Comment:
Is there a particular reason why you first visit the for body and then prepend the declarations?

Reply:
Yes because if we prepend `decls` before the visit, SSA transform will be applied to `decls` during the visit and redundant SSA variables will be created.

Reply:
Are you sure? `decls` should be SSA variables and thus no additional variables should be created.

Reply:
Ah you are right. But another issue is that non SSA identifier in declarations' RHS will be replaced with its value during traversal. ex. `let a_3 := a` becomes `let a_3 := a_3`.

Reply:
That should really only be done after the statement

Reply:
Sorry I don't understand. What thing should really only be done after which statement?